### ISSUE TITLE
Detection of non-deterministic renderings

### ISSUE THREAD

**Issue by hevrard** [google/amber#348]
Some shaders may lead to non-deterministic renderings, where running the same pipeline several times leads to different images. It would be great to have a way to detect this using Amber.

Unfortunately, calling `amber -i out.png nondet.shader_test` several times is not able to reproduce this non-determinism. It looks like we actually need to execute the same pipeline in the same "vulkan session" several times in a row to trigger such non-determinist renderings.

I can think of two ways to achieve this:
1. Add a special flag like `--check-nondet <n>` which tells to render the given test `<n>` times and report if the rendering is stable of not
2. Add a `dump` verb to the VkScript (or AmberScript?) DSL which enables to ask image rendering+dumping at particular points in the script, paving the way to dump the rendering of the same pipeline several times and do image comparison with external tools.

Does any of this sounds good? Any better idea on how to achieve nondet detection?

**Comment by dj2** [comment]
Questions to help me understand:

* Between runs, do we need to worry about clearing out the image buffer, or resetting state on things? Or can we just issue the draw call n times?

* Does the n in --check-nodet need to be variable or do you know a good value per shader?

* Do you want to run those shaders that way every time, or do you only check for determinism sometimes and have other expectations in there to check as well?

* Would something like:
```
RUN kRedPipeline DRAW_RECT POS 0 0 SIZE 256 256
COPY img_buffer TO tmp_buffer

REPEAT 10
  RUN kRedPipeline DRAW_RECT POS 0 0 SIZE 256 256
  EXPECT img_buffer EQ tmp_buffer
  COPY img_buffer TO tmp_buffer
END
```
Where img_buffer is setup to be a FRAMEBUFFER of kRedPipeline.

**Comment by hevrard** [comment]
> Questions to help me understand:
> 
> * Between runs, do we need to worry about clearing out the image buffer, or resetting state on things? Or can we just issue the draw call n times?

In graphicsfuzz vulkan harness, this is the DrawTest() function that is called several times with a different `png_filename` at each iteration:
https://github.com/google/graphicsfuzz/blob/aa0ce2ff6dab59329325d5df90ab7ae902112010/vulkan-worker/src/common/vulkan_worker.cc#L1604

The command buffer does not contain a clear command, see:
https://github.com/google/graphicsfuzz/blob/aa0ce2ff6dab59329325d5df90ab7ae902112010/vulkan-worker/src/common/vulkan_worker.cc#L1035

We use a swapchain, and acquire the next swapchain image, maybe this triggers some kind of clearing? (I would not think so).

> * Does the n in --check-nodet need to be variable or do you know a good value per shader?
We used to do 5 renderings, but it was cut down to 3 to make experiments go faster (given that we export to PNGs each time, each iteration has a significant cost)

It is nice for this argument to be variable as some nondet bugs need more iterations to reproduce than others (e.g. when the image is flickering approximately one time every 10 renderings). Also, when reporting a bug, we can set this number higher (e.g. 50 times) to make sure it will reproduce on the vendor side.

> * Do you want to run those shaders that way every time, or do you only check for determinism sometimes and have other expectations in there to check as well?

Detection of nondets are a best-effort because of their impredicable nature: if a nondet rendering bug needs 20 renderings to show up, we are likely to not detect it. Still, I think all nondet bugs found so far also trigger images which are not the ones expected with respect to the reference shader, so we are very likely to detect at least a wrong-image bug.

We typically want to run all experiments with a nondet checking across at least 3 renderings, more like 5 if the overhead is not too bad. Among thousands of results, it is nice to have some of them flagged as nondets because they are generally of higher priority to report to vendors.

> * Would something like:
> 
> ```
> RUN kRedPipeline DRAW_RECT POS 0 0 SIZE 256 256
> COPY img_buffer TO tmp_buffer
> 
> REPEAT 10
>   RUN kRedPipeline DRAW_RECT POS 0 0 SIZE 256 256
>   EXPECT img_buffer EQ tmp_buffer
>   COPY img_buffer TO tmp_buffer
> END
> ```

That looks good to me, we would not even need the COPY inside the loop (as all buffers are expected to be similar anyway).

**Comment by dj2** [comment]
As a first pass then, I think we need to add the following:

 * `COPY` command to copy the values from one buffer to another buffer
 * `REPEAT` command which will just take a value for now
 * `EXPECT` command to compare two buffers

That should at least let you get running. You can adjust the value in the script when sending to a vendor if you want. We could look at make it variable after the basic functionality is available.

Does that seem like it would work for you? Hopefully by doing the comparisons on memory and not writing the png files we can make it run fast enough to do the 5 comparisons but we'll have to see.

**Comment by hevrard** [comment]
Sounds good to me!
Let me know if I can help toward achieving this.

**Comment by dj2** [comment]
If you want to give `COPY` or `REPEAT` a go that would be cool. I'm working on adding simple `EXPECT` parsing at the moment which we can extend to buffers once the base is landed.

For both, need to add the parsing to AmberScript, update the docs/amber_script.md to list the command with a description.

For `COPY` implement a command which will make sure the Values are updated in the from buffer and then copy those values into the to buffer? We can put the restriction to start that these must be buffers defined with a format type (so will only be colour attachments) if that makes things easier to get running.

For `REPEAT`,  we'll have to do some magic in the parser to record the subsequent commands into a vector into the repeat command instead of the script. The execution should just loop over those commands ... I think?

**Comment by hevrard** [comment]
Thanks for the guidance, I will have a look tomorrow.

**Comment by hevrard** [comment]
I may not have time for this before a few days.

**Comment by dj2** [comment]
The `COPY` command will become a lot easier once my current chain of CLs land. All the data should be in the amber::Buffer object. You won't need to get anything back from the GPU, we do that automatically after a `draw` or `compute` call.

**Comment by dj2** [comment]
I think this is done now yea? We have COPY, EQ_BUFFER and REPEAT all landed.

**Comment by hevrard** [comment]
The shader used to test non-deterministic renderings needs some uniforms, does Amberscript enable to pass uniforms to shaders yet?

**Comment by dj2** [comment]
Uniforms should work. You BIND AS uniform instead of BIND AS storage. AmberScript does not add any padding to the buffer so you have to make sure you put in all the needed 0's for the stride.

If you run into issues, please file a bug. I'm currently working on getting push_constant's working in AmberScript.

**Comment by dj2** [comment]
I think this is all complete now. Please file new bugs if issues arise.
